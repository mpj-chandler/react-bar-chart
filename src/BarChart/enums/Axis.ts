enum Axis {
    XAxis,
    YAxis,
}

export default Axis;
