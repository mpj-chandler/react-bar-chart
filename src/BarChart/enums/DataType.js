"use strict";
exports.__esModule = true;
var DataType;
(function (DataType) {
    DataType["Named"] = "NAMED";
    DataType["DateIndexed"] = "DATE_INDEXED";
    DataType["Numeric"] = "NUMERIC";
    DataType["NonNullNumeric"] = "NON_NULL_NUMERIC";
})(DataType || (DataType = {}));
;
exports["default"] = DataType;
