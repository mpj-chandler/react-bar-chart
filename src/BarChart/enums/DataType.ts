enum DataType {
    Named = 'NAMED',
    DateIndexed = 'DATE_INDEXED',
    Numeric = 'NUMERIC',
    NonNullNumeric = 'NON_NULL_NUMERIC',
};

export default DataType;
