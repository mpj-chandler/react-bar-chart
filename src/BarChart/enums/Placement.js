"use strict";
exports.__esModule = true;
var Placement;
(function (Placement) {
    Placement[Placement["Aligned"] = 0] = "Aligned";
    Placement[Placement["Bucket"] = 1] = "Bucket";
})(Placement || (Placement = {}));
exports["default"] = Placement;
