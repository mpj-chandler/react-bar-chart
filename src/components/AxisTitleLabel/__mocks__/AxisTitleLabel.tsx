import React from "react";

const AxisTitleLabel: React.FC<any> = props => (
  <div {...props}>MockAxisTitleLabel {props.children}</div>
);

export default AxisTitleLabel;
