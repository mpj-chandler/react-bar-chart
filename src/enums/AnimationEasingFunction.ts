enum AnimationEasingType {
    Linear = 'linear',
    Elastic = 'elastic',
    InExpo = 'inExpo',
    None = 'none'
}

export default AnimationEasingType;