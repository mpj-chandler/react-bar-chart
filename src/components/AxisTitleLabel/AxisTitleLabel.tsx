import React, { useRef, useState, useEffect, RefObject } from "react";
import styles from "./AxisTitleLabel.scss";
import Axis from "../../enums/Axis";
import { number } from 'prop-types';

export interface AxisTitleLabelProps {
  label: string;
  titlePos?: {
      xPos: number
      yPos: number
  };
  axis: Axis;
}

const xAxisDefaults: Partial<AxisTitleLabelProps> = {
    titlePos: {
        xPos: 50,
        yPos: 95
    }
}

const yAxisDefaults: Partial<AxisTitleLabelProps> = {
    titlePos: {
        xPos: 5,
        yPos: 50
    }
};

const AxisTitleLabel: React.FC<AxisTitleLabelProps> = (props) => {

    const defaults: Partial<AxisTitleLabelProps> = props.axis === Axis.XAxis ? xAxisDefaults : yAxisDefaults;
    
    const normalisedProps: AxisTitleLabelProps = { ...props, titlePos: props.titlePos || defaults.titlePos };


  const ref = useRef();
  const [xTransform, setXTransform] = useState(0);
  const [yTransform, setYTransform] = useState(0);

  useEffect(() => {
    if (normalisedProps.axis === Axis.XAxis) {
      setXTransform(
        -0.5 *
          ((ref.current as unknown) as SVGTextContentElement).getBBox().width
      );

      return;
    }
    setXTransform(
      -1.25 *
        ((ref.current as unknown) as SVGTextContentElement).getBBox().width
    );
    setYTransform(
      0.25 *
        ((ref.current as unknown) as SVGTextContentElement).getBBox().height
    );
  }, [ref.current]);

  return (
    <text
      ref={(ref as unknown) as RefObject<SVGTextElement>}
      key={`tickText-${normalisedProps.label}`}
      className={styles.AxisTitleLabel}
      x={`${normalisedProps.titlePos.xPos}%`}
      y={`${normalisedProps.titlePos.yPos}%`}
      stroke={"black"}
      style={{ transform: `translate(${xTransform}px, ${yTransform}px)` }}
    >
      {normalisedProps.label}
    </text>
  );
};

export default AxisTitleLabel;
