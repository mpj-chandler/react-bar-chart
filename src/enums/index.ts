import Axis from './Axis';
import DataType from './DataType';
import { Placement } from './Placement';

export {
    Axis,
    DataType,
    Placement
}