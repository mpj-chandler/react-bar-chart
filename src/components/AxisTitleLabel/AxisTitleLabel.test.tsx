import AxisTitleLabel, { AxisTitleLabelProps } from './AxisTitleLabel';
import React, { ReactElement } from 'react';
import TestRenderer from 'react-test-renderer';
import { Axis } from '../../enums';

describe('AxisTitleLabel', () => {
    const label: string = "123456";

    const mockGetBBox = jest.fn().mockImplementation(() => {
        return {
            width: label.toString().length
        };
        });

        const nodeMock = (element: ReactElement) => {
        if (element.type === "text") {
            return {
            getBBox: () => {
                return {
                width: element.props.children[0].length
                };
            }
            };
        }

        return null;
        };
    describe("When mounted on an XAxis", () => {
    afterEach(() => {
        jest.resetAllMocks();
    });

    it("it renders consistently", () => {
        const component = TestRenderer.create(
        <AxisTitleLabel
            label={label}
            titlePos={{xPos: 0, yPos: 0}}
            axis={Axis.XAxis}
        />,
        { createNodeMock: nodeMock }
        );

        expect(component.toJSON()).toMatchSnapshot();
    });

    it("it adjusts to its text size", () => {
        let component = TestRenderer.create(
          <AxisTitleLabel
            label={label}
            titlePos={{ xPos: 0, yPos: 0 }}
            axis={Axis.XAxis}
          />,
          { createNodeMock: nodeMock }
        );

        const oldTransform = component.root.findByType("text").props.style
        .transform;

        component = TestRenderer.create(
          <AxisTitleLabel
            label={"123456789"}
            titlePos={{ xPos: 0, yPos: 0 }}
            axis={Axis.XAxis}
          />,
          { createNodeMock: nodeMock }
        );

        const newTransform = component.root.findByType("text").props.style
        .transform;

        expect(oldTransform).not.toStrictEqual(newTransform);
    });
    });

    describe("When mounted on a YAxis", () => {
        afterEach(() => {
            jest.resetAllMocks();
        });

        it("it renders consistently", () => {
            const component = TestRenderer.create(
              <AxisTitleLabel
                label={label}
                titlePos={{ xPos: 0, yPos: 0 }}
                axis={Axis.YAxis}
              />,
              { createNodeMock: nodeMock }
            );

            expect(component.toJSON()).toMatchSnapshot();
        });

        it("it adjusts to its text size", () => {
            let component = TestRenderer.create(
              <AxisTitleLabel
                label={label}
                titlePos={{ xPos: 0, yPos: 0 }}
                axis={Axis.YAxis}
              />,
              { createNodeMock: nodeMock }
            );

            const oldTransform = component.root.findByType("text").props.style
            .transform;

            component = TestRenderer.create(
              <AxisTitleLabel
                label={"123456789"}
                titlePos={{ xPos: 0, yPos: 0 }}
                axis={Axis.XAxis}
              />,
              { createNodeMock: nodeMock }
            );

            const newTransform = component.root.findByType("text").props.style
            .transform;

            expect(oldTransform).not.toStrictEqual(newTransform);
        });
    });
});