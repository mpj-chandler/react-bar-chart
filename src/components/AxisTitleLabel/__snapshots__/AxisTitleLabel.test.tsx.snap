// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AxisTitleLabel When mounted on a YAxis it renders consistently 1`] = `
<text
  className="AxisTitleLabel"
  stroke="black"
  style={
    Object {
      "transform": "translate(-1.25px, NaNpx)",
    }
  }
  x="0%"
  y="0%"
>
  123456
</text>
`;

exports[`AxisTitleLabel When mounted on an XAxis it renders consistently 1`] = `
<text
  className="AxisTitleLabel"
  stroke="black"
  style={
    Object {
      "transform": "translate(0px, 0px)",
    }
  }
  x="0%"
  y="0%"
>
  123456
</text>
`;
